source/gridfs/gridfs-spec.md

```diff
@@ -0,0 +1,1021 @@
+
+In the case of download_to_stream, drivers are not required to support partial file retrieval. If they choose to do so,
+drivers can support this operation by adding "start" and "end" to their supported options for download_to_stream. These
+values represent non-negative byte offsets from the beginning of the file. When "start" and "end" are specified, drivers
+return the bytes of the file in `[start, end)`. If "start" and "end" are equal no data is returned.
+
+If either "start" or "end" is invalid, drivers MUST raise an error. These values are considered invalid if they are
+negative, greater than the file length, or if "start" is greater than "end".
+
+When performing partial reads, drivers SHOULD use the file"s "chunkSize" to calculate which chunks contain the desired
+section and avoid reading unneeded documents from the "chunks" collection.
+
+### Renaming stored files
+
+```javascript
+class GridFSBucket {
+
+  /**
+   * Renames the stored file with the specified @id.
+   */
+  void rename(TFileId id, string new_filename);
+
+}
+```
+
+Sets the filename field in the stored file"s files collection document to the new filename.
+
+**Implementation details:**
+
+Drivers construct and execute an update_one command on the files collection using `{ _id: @id }` as the filter and
+`{ $set : { filename : "new_filename" } }` as the update parameter.
+
+To rename multiple revisions of the same filename, users must retrieve the full list of files collection documents for a
+given filename and execute "rename" on each corresponding `_id`.
+
+If there is no file with the given id, drivers MUST raise an error.
+
+### Dropping an entire GridFS bucket
+
+```javascript
+class GridFSBucket {
+
+  /**
+   * Drops the files and chunks collections associated with
+   * this bucket.
+   */
+  void drop();
+
+}
+```
+
+This method drops the files and chunks collections associated with this GridFS bucket.
+
+Drivers should drop the files collection first, and then the chunks collection.
+
+## Test Plan
+
+TBD
+
+## Motivation for Change
+
+The [existing GridFS documentation](https://www.mongodb.com/docs/manual/reference/gridfs/) is only concerned with the
+underlying data model for this feature, and does not specify what basic set of features an implementation of GridFS
+should or should not provide. As a result, GridFS is currently implemented across drivers, but with varying APIs,
+features, and behavior guarantees. Current implementations also may not conform to the existing documentation.
+
+This spec documents minimal operations required by all drivers offering GridFS support, along with optional features
+that drivers may choose to support. This spec is also explicit about what features/behaviors of GridFS are not specified
+and should not be supported. Additionally, this spec validates and clarifies the existing data model, deprecating fields
+that are undesirable or incorrect.
+
+## Design Rationale
+
+Why is the default chunk size 255 KiB?\
```

Comment:
Should the questions be bolded for readability, as they were in the RST? I forget if you did this in previous conversions.

Reply:
Done